Draft for Schema

ViktorLindgren95 opened the pull request:
Added the new schema
Added Error handling
Refactoring

This is a draft pull request i just want you to look at the changes ive made here and then we can discuss it in our collaboration meeting.

Commit:
Added the new schema
Added Error handling
Refactoring

felixlut commented:
Proposal:
Switch from a GitHub repository based approach to instead base the onboarding on a Team instead. This will instead make it so that a team is an owner of multiple GitHub repositories, with corresponding Sonarqube `Projects` which are all part of a SonarQube `Portfolio`:

JSON:
```json
{
    "Name": "My-team",
    "Visibility": "Public",
    "Users": [
        {
            "Username": "User1",
            "Role": "Admin"
        },
        {
            "Username": "User2",
            "Role": "Write"
        }
    ],
    "Components": {
        "GitHub": {
            "Repositories": [
                {
                    "Name": "Org/Repo",
                    "Topics": [
                        "Topic1",
                        "Topic2"
                    ],
                    "Status": {
                        "State": "state",
                        "ReconsiledAt": "Time"
                    },
                    "Outside-Collaborators": [
                        
                    ]
                }
            ]
        },
        "Sonarqube": {
            "Portfolio": {
                "Name": "Portfolio1",
                "Status": {
                    "State": "state",
                    "ReconsiledAt": "Time"
                },
                "Projects": [
                    {
                        "Name": "MySQProj",
                        "Key": "MySQProj",
                        "Qualifier": "TKR/VW/APP",
                        "Visibility": "Public/private",
                        "Status": {
                            "State": "Status",
                            "ReconsiledAt": "Time"
                        }            
                    }
                ]
            }
        }
    }
}
```

Golang struct:
```go
type Team struct {
	Name       string     `json:"Name"`
	Visibility string     `json:"Visibility"`
	Users      []Users    `json:"Users"`
	Components Components `json:"Components"`
}
type Users struct {
	Username string `json:"Username"`
	Role     string `json:"Role"`
}

type Components struct {
	GitHub    GitHub    `json:"GitHub"`
	Sonarqube Sonarqube `json:"Sonarqube"`
}

type GitHub struct {
	Repositories []Repositories `json:"Repositories"`
}
type Repositories struct {
	Name                 string        `json:"Name"`
	Topics               []string      `json:"Topics"`
	Status               Status        `json:"Status"`
	OutsideCollaborators []interface{} `json:"Outside-Collaborators"`
}


type Sonarqube struct {
	Portfolio Portfolio `json:"Portfolio"`
}
type Portfolio struct {
	Name     string     `json:"Name"`
	Status   Status     `json:"Status"`
	Projects []Projects `json:"Projects"`
}
type Projects struct {
	Name       string `json:"Name"`
	Key        string `json:"Key"`
	Qualifier  string `json:"Qualifier"`
	Visibility string `json:"Visibility"`
	Status     Status `json:"Status"`
}


type Status struct {
	State        string `json:"State"`
	ReconsiledAt string `json:"ReconsiledAt"`
}
```

ViktorLindgren95 commented:
> ```json
> Role
> ```

Alright but then we still need to make it so users can add a repo and then just attach it to a team right? Like the mother of object is a team object but users can still do like " add repo -> team -> my team " in some kind of interface

But it makes for a better structure i believe. but it also becomes a monster of a file so maybe its not good

felixlut committed:
add extraMemebers and team field to github repo

Commit:
Sorted out our import cycle
Implemented the new structure
Added functions for team

ViktorLindgren95 committed:
Commented out sonarqube for local testing moving forward.

Testing was sucessfull for creation. the function that calls for rename is broken and needs to be fixed so we can rename and edit the repository to test the other function :).

ViktorLindgren95 committed:
Solved the parsing function by saving the body inside some bytes :)

Commit:
Added Env handling for most of our environment secrets. also some tests

Commit:
Added readme
Made the decision to make conversions on the function level

Commit:
According to tests my idea didnt work.

So conversion for single use is done on function level
reused things are fetched in Init of the usage function. right now that is createSourcehook

Commit:
According to tests my idea didnt work.

So conversion for single use is done on function level
reused things are fetched in Init of the usage function. right now that is createSourcehook

Commit:
Suggestion for breakout

Also we kinda need to discuss properly with the environment variables haha

Commit:
Broke out the function. Waiting for merge until further work on the breakout

Commit:
Broke out the function. Waiting for merge until further work on the breakout

Commit:
Added Branch Rule function
Added Workflow function

ViktorLindgren95 committed:
Apparently just adding the file and reading it works for pushing here.
Pretty neat honestly. We could move this to a configmap or something eventually (see how grafana does this) but for now this works well

ViktorLindgren95 committed:
Merge branch 'schema' into schema-env

ViktorLindgren95 committed:
Merge pull request #3 from FridayCommit/schema-env

Added some env variables handling

ViktorLindgren95 committed:
Merge branch 'add_rules' into schema
